contracts/0.4.24/oracle/LidoOracle.sol

```diff
@@ -380,60 +459,76 @@ contract LidoOracle is ILidoOracle, AragonApp {
      * @return isQuorum - true, when quorum is reached, false otherwise
      * @return modeReport - valid mode-value report when quorum is reached, 0-data otherwise
      */
-    function _getQuorumReport(uint256 _epochId) internal view returns (bool isQuorum, Report memory modeReport) {
-        uint256 mask = gatheredEpochData[_epochId].reportsBitMask;
-        uint256 popcnt = mask.popcnt();
-        if (popcnt < getQuorum())
-            return (false, Report({beaconBalance: 0, beaconValidators: 0}));
+    function _getQuorumReport() internal view returns (bool isQuorum, Report memory modeReport) {
+        uint256 quorum = getQuorum();
 
-        assert(0 != popcnt && popcnt <= members.length);
+        // check most frequent cases: all reports are the same or no reports yet
+        if (gatheredReportKinds.length == 1) {
+            return (gatheredReportKinds[0].count >= quorum, gatheredReportKinds[0].report);
+        } else if (gatheredReportKinds.length == 0) {
+            return (false, Report({beaconBalance: 0, beaconValidators: 0}));
+        }
 
-        // pack current gatheredEpochData mapping to uint256 array
-        uint256[] memory data = new uint256[](popcnt);
-        uint256 i = 0;
-        uint256 membersLength = members.length;
-        for (uint256 index = 0; index < membersLength; ++index) {
-            if (mask.getBit(index)) {
-                data[i++] = reportToUint256(gatheredEpochData[_epochId].reports[index]);
+        // if more than 2 kind of reports exist, choose the most frequent
+        uint256 maxi = 0;
+        for (uint256 i = 1; i < gatheredReportKinds.length; ++i) {
+            if (gatheredReportKinds[i].count > gatheredReportKinds[maxi].count) maxi = i;
+        }
+        for (i = 0; i < gatheredReportKinds.length; ++i) {
```

Comment:
I think we can improve this algorithm and merge that two loops in one. What do you think?

Reply:
Yes, implemented: https://github.com/lidofinance/lido-dao/pull/313/commits/0c3110f48f59361caefafddf0401c18a30dbf60d. Although this code is run extremally rarely so Vasiliy initially asked me to make it "just works" :)